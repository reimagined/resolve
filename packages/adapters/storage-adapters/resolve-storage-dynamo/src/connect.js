import { apiVersion } from './constants'

const connect = async (pool, { DynamoDB, checkTableExists, ...helpers }) => {
  const {
    tableName,
    billingMode = 'PAY_PER_REQUEST',
    readCapacityUnits = 5,
    writeCapacityUnits = 5,
    ...connectionOptions
  } = pool.config

  const options = {
    ...connectionOptions,
    apiVersion
  }

  const database = new DynamoDB(options)
  const documentClient = new DynamoDB.DocumentClient(options)

  Object.assign(pool, helpers, {
    database,
    documentClient,
    checkTableExists,
    tableName,
    billingMode,
    readCapacityUnits,
    writeCapacityUnits
  })
}

export default connect
