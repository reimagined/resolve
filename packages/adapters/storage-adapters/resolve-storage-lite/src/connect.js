const connect = async (pool, { NeDB, promiseInvoke }) => {
  const { pathToFile, ...initOptions } = pool.config

  const database = new NeDB({
    ...(pathToFile != null ? { filename: pathToFile } : { inMemoryOnly: true }),
    ...initOptions
  })

  Object.assign(pool, {
    promiseInvoke,
    database
  })
}

export default connect
