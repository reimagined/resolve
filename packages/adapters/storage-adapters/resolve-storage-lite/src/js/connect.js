import getLog from './get-log'

const coerceEmptyString = obj =>
  (obj != null && obj.constructor !== String) || obj == null ? 'default' : obj

const connectEventStore = async (pool, { sqlite, tmp, os, fs }) => {
  const log = getLog('connectEventStore')

  log.debug(`connecting to events database`)
  const { escape } = pool
  let { databaseFile, tableName, ...initOptions } = pool.config

  databaseFile = coerceEmptyString(databaseFile)
  tableName = coerceEmptyString(tableName)

  log.verbose(`databaseFile: ${databaseFile}`)
  log.verbose(`tableName: ${tableName}`)

  let connector
  if (databaseFile === ':memory:') {
    log.debug(`using memory connector`)
    if (process.env.RESOLVE_LAUNCH_ID != null) {
      const tmpName = `${os.tmpdir()}/storage-${+process.env
        .RESOLVE_LAUNCH_ID}.db`
      const removeCallback = () => {
        if (fs.existsSync(tmpName)) {
          fs.unlinkSync(tmpName)
        }
      }

      if (!fs.existsSync(tmpName)) {
        fs.writeFileSync(tmpName, '')
        process.on('SIGINT', removeCallback)
        process.on('SIGTERM', removeCallback)
        process.on('beforeExit', removeCallback)
        process.on('exit', removeCallback)
      }

      pool.memoryStore = {
        name: tmpName,
        drop: removeCallback
      }
    } else {
      const temporaryFile = tmp.fileSync()
      pool.memoryStore = {
        name: temporaryFile.name,
        drop: temporaryFile.removeCallback.bind(temporaryFile)
      }
    }

    connector = sqlite.open.bind(sqlite, pool.memoryStore.name)
  } else {
    log.debug(`using disk file connector`)
    connector = sqlite.open.bind(sqlite, databaseFile)
  }

  log.debug(`connecting`)
  const database = await connector()

  log.debug(`adjusting connection`)

  log.verbose(`PRAGMA busy_timeout=1000000`)
  await database.exec(`PRAGMA busy_timeout=1000000`)

  log.verbose(`PRAGMA encoding=${escape('UTF-8')}`)
  await database.exec(`PRAGMA encoding=${escape('UTF-8')}`)

  log.verbose(`PRAGMA synchronous=EXTRA`)
  await database.exec(`PRAGMA synchronous=EXTRA`)

  if (databaseFile === ':memory:') {
    log.verbose(`PRAGMA journal_mode=MEMORY`)
    await database.exec(`PRAGMA journal_mode=MEMORY`)
  } else {
    log.verbose(`PRAGMA journal_mode=DELETE`)
    await database.exec(`PRAGMA journal_mode=DELETE`)
  }

  Object.assign(pool, {
    database,
    tableName,
    initOptions
  })

  log.debug(`events store database connected successfully`)
}

export default connectEventStore
