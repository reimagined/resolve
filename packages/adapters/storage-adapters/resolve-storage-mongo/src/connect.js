const connect = async (pool, { MongoClient }) => {
  const { url, collectionName, ...connectionOptions } = pool.config

  const client = await MongoClient.connect(url, {
    ...connectionOptions,
    useNewUrlParser: true
  })
  const database = await client.db()
  const collection = await database.collection(collectionName)

  Object.assign(pool, {
    collectionName,
    client,
    database,
    collection
  })
}

export default connect
