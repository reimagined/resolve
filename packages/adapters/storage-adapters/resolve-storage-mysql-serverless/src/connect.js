const coercer = ({ intValue, stringValue, bigIntValue, ...rest }) => {
  if (intValue != null) {
    return Number(intValue)
  } else if (bigIntValue != null) {
    return Number(bigIntValue)
  } else if (stringValue != null) {
    return String(stringValue)
  } else {
    throw new Error(`Unknown type ${JSON.stringify(rest)}`)
  }
}

const executeSql = async (pool, sql) => {
  const result = await pool.rdsDataService
    .executeSql({
      awsSecretStoreArn: pool.config.awsSecretStoreArn,
      dbClusterOrInstanceArn: pool.config.dbClusterOrInstanceArn,
      database: pool.config.databaseName,
      sqlStatements: sql
    })
    .promise()

  if (
    !Array.isArray(result.sqlStatementResults) ||
    result.sqlStatementResults.length !== 1
  ) {
    return null
  }

  const firstResult = result.sqlStatementResults[0]

  if (firstResult == null || firstResult.resultFrame == null) {
    return null
  }

  const {
    resultFrame: { records, resultSetMetadata }
  } = firstResult

  if (!Array.isArray(records) || resultSetMetadata == null) {
    return null
  }

  const { columnCount, columnMetadata } = resultSetMetadata

  const rows = []
  for (const { values } of records) {
    const row = {}
    for (let i = 0; i < columnCount; i++) {
      row[columnMetadata[i].name] = coercer(values[i])
    }
    rows.push(row)
  }

  return rows
}

const connect = async (
  pool,
  { RDSDataService, escapeId, escapeUnicode, escape }
) => {
  const rdsDataService = new RDSDataService({ region: pool.config.region })

  Object.assign(pool, {
    tableName: pool.config.tableName,
    rdsDataService,
    executeSql: executeSql.bind(null, pool),
    escapeUnicode,
    escapeId,
    escape,
    resourceOptions: pool.config.resourceOptions
  })
}

export default connect
