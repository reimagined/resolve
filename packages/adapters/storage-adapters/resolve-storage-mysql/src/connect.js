const connect = async (pool, { MySQL, escapeId, escape }) => {
  const { tableName, database, ...connectionOptions } = pool.config

  const connection = await MySQL.createConnection({
    ...connectionOptions,
    database,
    multipleStatements: true
  })

  Object.assign(pool, {
    connection,
    tableName,
    escapeId,
    escape,
    database
  })
}

export default connect
