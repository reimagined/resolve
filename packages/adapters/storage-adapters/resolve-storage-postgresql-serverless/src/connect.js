const connect = async (
  pool,
  {
    RDSDataService,
    escapeId,
    escape,
    fullJitter,
    randRange,
    executeStatement,
    coercer
  }
) => {
  const rdsDataService = new RDSDataService({ region: pool.config.region })

  Object.assign(pool, {
    tableName: pool.config.tableName,
    rdsDataService,
    fullJitter,
    randRange,
    coercer,
    executeStatement: executeStatement.bind(null, pool),
    databaseName: pool.config.databaseName,
    escapeId,
    escape
  })
}

export default connect
