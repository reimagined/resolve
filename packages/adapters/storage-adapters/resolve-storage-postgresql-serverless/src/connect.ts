import getLog from './js/get-log'
import { AdapterPool, AdapterSpecific } from './types'

const connect = async (
  pool: AdapterPool,
  specific: AdapterSpecific
): Promise<any> => {
  const log = getLog('connect')
  log.debug('configuring RDS data service client')

  const {
    RDSDataService,
    escapeId,
    escape,
    fullJitter,
    executeStatement,
    coercer
  } = specific

  const {
    dbClusterOrInstanceArn,
    awsSecretStoreArn,
    databaseName,
    tableName,
    secretsTableName,
    ...rdsConfig
  } = pool.config

  const rdsDataService = new RDSDataService(rdsConfig)

  Object.assign(pool, {
    rdsDataService,
    dbClusterOrInstanceArn,
    awsSecretStoreArn,
    databaseName,
    tableName,
    secretsTableName,
    fullJitter,
    coercer,
    executeStatement: executeStatement.bind(null, pool),
    escapeId,
    escape
  })

  log.debug('RDS data service client configured')
}

export default connect
