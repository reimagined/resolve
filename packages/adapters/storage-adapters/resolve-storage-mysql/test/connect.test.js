import sinon from 'sinon'

import connect from '../src/connect'

test('connect should connect to mysql database', async () => {
  const connection = {
    execute: sinon.stub().callsFake(async () => null)
  }
  const MySQL = {
    createConnection: sinon.stub().callsFake(async () => connection)
  }

  const escapeId = value => `@ESCAPED[${value}]`
  const escape = 'escape'

  const pool = {
    config: {
      host: 'host',
      port: 'port',
      user: 'user',
      password: 'password',
      database: 'database',
      tableName: 'tableName'
    }
  }

  await connect(
    pool,
    { MySQL, escapeId, escape }
  )

  expect(MySQL.createConnection.callCount).toEqual(1)
  expect(MySQL.createConnection.firstCall.args).toMatchSnapshot()

  expect(pool.tableName).toEqual(pool.config.tableName)
  expect(pool.escapeId).toEqual(escapeId)
  expect(pool.escape).toEqual(escape)
  expect(pool.connection).toEqual(connection)
})
