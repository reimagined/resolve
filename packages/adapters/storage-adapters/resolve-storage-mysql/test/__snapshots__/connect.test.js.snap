// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`connect should connect to mysql database 1`] = `
Array [
  Object {
    "database": "database",
    "host": "host",
    "multipleStatements": true,
    "password": "password",
    "port": "port",
    "user": "user",
  },
]
`;
